Store and expose "links" of entries

tekin opened the pull request:
To make it possible to infer relationships of entries in the content register, we want to expose "links". This updates the JSON representation of entries to include a partially-expanded set of links, currently only showing the `content_id`. Eventually, the expanded form will include details such as the title, format, base_path, etc, but that is a much larger piece of work that requires de-referencing of content_ids against other entries in content-register. For now, it is enough for us to partially expand the links, providing a future-proof data-structure that can be expanded later to include the other fields.

Trello: https://trello.com/c/dD6bkH5j/180-infer-programme-to-policy-relationship-when-tagging-editions

tekin committed:
Tweak rails_helper.rb ordering

Requiring rspec/rails before spec_helper.rb stops app preloaders like zeus
blowing up when trying to load the spec environment.

tekin committed:
Add serialised "links" to Entry model

We're going to store the "links" hash of content items with their entries so
that that information can be included to clients of the content register API.

This is an interim solution as the finding things team have upcoming work to
enable the checking of taggings across models, which will inevitably mean
storing a normalised representation of the related "links".

tekin committed:
Include links in the JSON representation of Entry

This also reworks the spec to be explicit about exactly which attributes we
expect to be public on the model, otherwise it's too much like testing
implementation and also because changes to Entry::PUBLIC_ATTRIBUTES would not
result in failing specs, even though it would break behaviour.

tekin committed:
Record links received over the message queue

Ensure that links are stored and updated based on the data received on the
message queue.

tekin committed:
Correct format of links from message queue

The representation of content items received on the message queue includes links
in their unexpanded form, i.e. arrays of content_ids.

tekin committed:
Use standard :id symbol to make routing simpler

By using conventional :id symbol to represent the identifier for an entry in
the routing, we can leverage the standard RESTful URL helper behaviour and don't
have to explicitly specify the key when using routing helpers.

tekin committed:
Make integration tests more robust

Specifically, this makes the spec robust against changes to the JSON
representation of an entry, i.e. they won't break every time the JSON
representation of an Entry changes.

tekin committed:
Expand out "links" on Entry JSON representation

We want entries to include the expanded form of their links. Eventually, these
will include details such as the title, format, base_path, etc, but that is a
much larger piece of work that requires de-referencing content_ids against other
entries in content-register. For now, it is enough for us to partially expand
the links, only including the "content_id". This is a future-proof
data-structure that can be expanded later to include the other fields.

alext commented on spec/models/entry_spec.rb:
```diff
@@ -23,13 +23,23 @@
+
+    entry.links = links
+    entry.save!
+    expect(entry.links).to eq(links)
```
It is worth doing a fresh find here to validate that the data survived the round-trip to the database.

alext commented:
One small comment, otherwise looks good to me.

tekin committed:
Update docs to reflect changes to entry format

tekin committed:
Validate incoming "links" attribute

Validating that the incoming "links" are in the correct format will prevent
errors being raised when transforming them on the way back out.

tekin commented:
Updated based on comment. Also updated the docs to reflect the new input/output formats, and added validations on the incoming "links" to ensure things don't break if entries are submitted with an unexpected format.

heathd commented on spec/models/entry_spec.rb:
```diff
@@ -23,13 +23,43 @@
   end
 
+  context "the links hash attribute" do
+      it "stores a serialized links hash" do
```
some excess indentation here

tekin replied:
Thanks, fixed.

tekin committed:
Fix indentation

heathd commented:
:+1: all looks good

I feel slightly nervous about the lack of tests between content store and content register (verification that the format sent on the message queue matches what content register expects), but we don't really have any infrastructure in place to deal with that, and I don't have a good suggestion about how to do that.

It's good that you've got the stronger validations on the input.
